Create new PHPUnit assertions for the WebTestCase

Pierstoval opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| License       | MIT
| Doc PR        | ~

I got inspirations from Laravel Dusk's assertions to bring new assertion capabilities to the `WebTestCase` in order to ease functional testing.

WDYT?

Edit: Here's the list of the added assertions, as a summary of this PR:

* `assertResponseIsSuccessful`
* `assertHttpCodeEquals`
* `assertResponseHasHeader`
* `assertResponseNotHasHeader`
* `assertResponseHeaderEquals`
* `assertResponseHeaderNotEquals`
* `assertResponseRedirects`
* `assertPageTitleEquals`
* `assertPageTitleContains`
* `assertClientHasCookie`
* `assertClientNotHasCookie`
* `assertClientCookieValueEquals`
* `assertClientRawCookieValueEquals`
* `assertResponseHasCookie`
* `assertResponseNotHasCookie`
* `assertResponseCookieValueEquals`
* `assertResponseCookieValueNotEquals`
* `assertSelectorExists`
* `assertSelectorNotExists`
* `assertSelectorContainsText`
* `assertSelectorNotContainsText`
* `assertInputValueEquals`
* `assertInputValueNotEquals`
* `assertRouteEquals`
* `assertRequestAttributeValueEquals`

Pierstoval committed:
Create new PHPUnit assertions for the WebTestCase

stof commented on src/Symfony/Bundle/FrameworkBundle/Test/WebTestCase.php:
```diff
@@ -41,6 +43,6 @@ protected static function createClient(array $options = [], array $server = [])
         $client->setServerParameters($server);
 
-        return $client;
+        return static::$client = $client;
```
if you store the client in the property, you must ensure you clean it on teardown so that we don't keep unwanted references to it.

Pierstoval replied:
Yep, done! Thanks :+1:

Pierstoval commented on src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php:
```diff
@@ -0,0 +1,242 @@
+
+        $response = static::$client->getResponse();
+
+        Assert::assertSame(
```
Same vs equals is something that is specific to phpunit (and also to Twig), in PHP we make the difference by talking about strict or non-strict equality, so I think it's more convenient to stick to a more common approach of naming equality with "equals", WDYT?

localheinz replied:
I believe in intention revealing interfaces.

An invocation of this assertion will fail when `$expectedCode` is an `int` casted to a `string`, whereas - for example - an invokation of `assertEquals()` will pass.

mhujer commented on src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php:
```diff
@@ -0,0 +1,292 @@
+        );
+    }
+
+    public static function assertResponseRedirects(string $expectedLocation = null): void
```
@Pierstoval What about adding an expected redirect code (301 or 302)? I have it in my own `WebTestCase.php` - https://gist.github.com/mhujer/6754c021e0913e4b11c1788ac06ca768

sstok commented on src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php:
```diff
@@ -0,0 +1,292 @@
+
+        $cookie = static::$client->getCookieJar()->get($name, $path, $domain);
+
+        Assert::assertNotNull(
```
Maybe return which cookies did exist? In case someone made a type.

Pierstoval replied:
Not sure, if the developer made a typo they will see it in the error message, since it's added in the assertion exception message

sstok commented on src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php:
```diff
@@ -0,0 +1,292 @@
+        Assert::assertSame(
+            $expectedValue,
+            $value = $cookie->getRawValue(),
+            "Cookie value [$expectedValue] does not match actual value [$value]."
```
Mention the cookie name?

Pierstoval replied:
Yep, thanks for the catch!

Pierstoval commented on src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php:
```diff
@@ -0,0 +1,292 @@
+
+        $titleNode = static::$client->getCrawler()->filter('title');
+
+        Assert::assertGreaterThan(0, $titleNode->count(), 'There is no <title> tag in the current page to look for.');
```
You're right, I'll fix this

mhujer commented on src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php:
```diff
@@ -0,0 +1,296 @@
+        );
+    }
+
+    public static function assertResponseRedirects(int $expectedCode = null, string $expectedLocation = null): void
```
What about swapping the parameters? Having the `$expectedLocation` first will allow someone just to check that when the redirect code is not important.

mhujer commented on src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php:
```diff
@@ -0,0 +1,296 @@
+            "Actual response code [{$response->getStatusCode()}] does not match any valid HTTP redirection code."
+        );
+
+        if ($expectedCode) {
```
```suggestion
        if ($expectedCode !== null) {
```

mhujer commented on src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php:
```diff
@@ -0,0 +1,296 @@
+            static::assertHttpCodeEquals($expectedCode);
+        }
+
+        if ($expectedLocation) {
```
```suggestion
        if ($expectedLocation !== null) {
```

Otherwise it will silently skip the check if someone accidentally puts `''` as an `$expectedLocation`.

Pierstoval replied:
A location that equals `""` is not something valid, so this should never happen

mhujer replied:
@Pierstoval It should never happen, but it can. But in case I accidentally put there `''` as expectedLocation (e.g. when i will have a list of URLs in the testcase and I'm iterating through them) I would prefer to know it immediately instead of letting it fail silently.

Pierstoval replied:
Well, I can change this so any non-null may fail in this case, seems fair

mhujer commented on src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php:
```diff
@@ -0,0 +1,296 @@
+
+        $titleNode = static::$client->getCrawler()->filter('title');
+
+        Assert::assertSame(1, $titleNode->count(), 'There is no <title> tag in the current page to look for.');
```
The message may be misleading if there are two `<title>` tags. Maybe have to assertions? First check that there is at least one (with greaterThan and message `'There is no <title> tag in the current page to look for.'` and second `Assert::assertSame(1, $titleNode->count(), 'There must be exactly one <title> tag in the page.');`

mhujer replied:
or maybe an `assertCount` will work better?

dunglas reviewed (approved):
Cool!

localheinz commented on src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php:
```diff
@@ -0,0 +1,296 @@
+    /** @var Client|null */
+    protected static $client;
+
+    public static function assertResponseIsSuccessful(): void
```
Do these methods really need `public` visibility, or will `protected` suffice?

Pierstoval replied:
I prefer to keep them public like the `Assert` class does, in case someone wants to test it or in case someone doesn't want to use the trait.

localheinz replied:
But how would you be able to test the trait _without_ using it?

Pierstoval replied:
Precisely by making its methods static and public 😄 
I just copied the behavior of the `Assert` class, that's all, and I think it's fine as-is

localheinz replied:
> Precisely by making its methods static and public 😄

Can you give a concrete example?

I don't think there is a need to unnecessarily use a visibility that is not needed.

Pierstoval replied:
I have no special argument, I just think it's fine to be able to use it without using the trait

localheinz commented on src/Symfony/Bundle/FrameworkBundle/Test/WebTestCase.php:
```diff
@@ -21,6 +21,15 @@
 {
+    use WebTestAssertions;
+
+    public function tearDown()
```
I believe `protected` visibility should suffice here. 

What do you think?

Pierstoval replied:
Yep, didn't remember that the original method is `protected`, thanks 👍

localheinz commented on src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php:
```diff
@@ -0,0 +1,296 @@
+
+        Assert::assertTrue(
+            $response->isRedirect(),
+            "Actual response code [{$response->getStatusCode()}] does not match any valid HTTP redirection code."
```
What do you think about streamlining these messages?

See above in `assertResponseIsSuccessful`: 

```
"Response was expected to be successful, but actual HTTP code is {$response->getStatusCode()}."`
```

Perhaps here we use

```
"Response was expected to be a redirect, but actual HTTP code is {$response->getStatusCode()}."`
```

Pierstoval replied:
Thanks for the suggestion, I'll add it 👍

localheinz commented on src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php:
```diff
@@ -0,0 +1,296 @@
+        Assert::assertContains(
+            $expectedTitle,
+            \trim($title = $titleNode->text()),
+            "Expected title [$expectedTitle] does not equal actual title [$title]."
```
Should this be

```
"Expected title [$expectedTitle] does not contain actual title [$title]."
```

?

Pierstoval replied:
Will fix, thanks 👍

localheinz commented on src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php:
```diff
@@ -0,0 +1,296 @@
+
+        Assert::assertGreaterThan(0, $nodes->count(), "Selector [$selector] does not resolve to any node.");
+
+        Assert::assertContains($text, $nodes->text(), "Selector [$selector] does not contain text [$text].");
```
Should this be

```php
Assert::assertNotContains($text, $nodes->text(), "Selector [$selector] contains text [$text].");
```

?

Pierstoval replied:
Fixed, and I also reworded to `"Selector [$selector] was expected to not contain text [$text]."` instead, I think it is more explanatory 👍

localheinz commented on src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php:
```diff
@@ -0,0 +1,296 @@
+        Assert::assertNotEquals(
+            $expectedValue,
+            $value = $inputNode->getNode(0)->getAttribute('value'),
+            "Expected value [$value] for the [$fieldName] input does not equal the actual value [$value]."
```
Should this be

```
"Expected value [$value] for the [$fieldName] input equals the actual value [$value]."
```

XuruDragon commented:
Thank you ! so tired of always recoding the same bits of tests... This will greatly increase testing :)

Pierstoval commented:
Thanks!

Friendly update to the reviewers: I added a slight summary of the assertions in the PR description so we can directly see what's added, for easier review.

Waiting for more reviews 👍

ro0NL commented on src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php:
```diff
@@ -0,0 +1,413 @@
+
+        Assert::assertSame(
+            $expectedValue,
+            $value = $response->headers->get($headerName),
```
i think we should test all header values here, i.e. `get($headerName, null, true)`

Pierstoval replied:
Thanks, solved!

ro0NL commented on src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php:
```diff
@@ -0,0 +1,413 @@
+
+        Assert::assertNotSame(
+            $expectedValue,
+            $value = $response->headers->get($headerName),
```
and here as well

renanbr commented on src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php:
```diff
@@ -0,0 +1,413 @@
+        );
+        Assert::assertSame(
+            $expectedValue,
+            $value = $cookie->getRawValue(),
```
```suggestion
            $value = $cookie->getValue(),
```

Pierstoval replied:
Fixed 👍

renanbr commented on src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php:
```diff
@@ -0,0 +1,413 @@
+
+        Assert::assertGreaterThan(0, $nodes->count(), "Selector [$selector] does not resolve to any node.");
+
+        Assert::assertContains($text, $nodes->text(), "Selector [$selector] does not contain text [$text].");
```
Does it check against the first node of the list intentionally?
If not, is there a way to assert against all nodes?

same for `assertSelectorNotContainsText`

Pierstoval replied:
Yep, I think checking _all_ nodes might be a bit cumbersome, and it probably corresponds to non-common behavior

XuruDragon commented on src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php:
```diff
@@ -0,0 +1,413 @@
+        if (!static::$client->getResponse()) {
+            static::fail('A client must have an HTTP Response to make assertions.');
+        }
+    }
```
Maybe your `checkResponseAvailable()` could be more usefull by return the response than void?

As you do `static::$client->getResponse()` every time after `checkResponseAvailable()` call

Pierstoval replied:
I removed the "check" methods and just use getters 👍

XuruDragon commented on src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php:
```diff
@@ -0,0 +1,413 @@
+        if (!static::$client->getRequest()) {
+            static::fail('A client must have an HTTP Request to make assertions.');
+        }
+    }
```
Same than above but return the request

ro0NL commented on src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php:
```diff
@@ -0,0 +1,382 @@
+        );
+    }
+
+    public static function assertResponseHeaderEquals(string $headerName, string $expectedValue): void
```
`array $expectedValue`

Pierstoval replied:
I'll go for `mixed` instead then, it can be different types

localheinz reviewed (changes requested):
Why are there no tests?

stof commented on src/Symfony/Bundle/FrameworkBundle/Test/WebTestCase.php:
```diff
@@ -21,6 +21,15 @@
+    use WebTestAssertions;
+
+    protected function tearDown()
+    {
```
you need to also call the parent method, otherwise the kernel is not shutdown properly

Pierstoval replied:
Fixed in 86fdbb8

stof commented on src/Symfony/Bundle/FrameworkBundle/Test/WebTestCase.php:
```diff
@@ -21,6 +21,15 @@
 {
+    use WebTestAssertions;
+
+    protected function tearDown()
```
this will cause issues with PHPUnit 8, while KernelTestCase was adapted to support it.

Pierstoval replied:
What would you recommend then?
Would it be ok to create another trait, like `RemoveClientOnShutdownTrait` or similar?

nicolas-grekas replied:
see #30474

Pierstoval replied:
Awesome! Thanks for the link, I'll do this 👍

Pierstoval replied:
Fixed in 86fdbb8

Pierstoval commented:
@localheinz:
> Why are there no tests?

I guess because the `KernelTestCase` is not tested?

Pierstoval commented:
Friendly ping to @dunglas and @stof 

I fixed the compatibility with PHPUnit8 by using the same technique as https://github.com/symfony/symfony/pull/30474/files 

However, the aforementioned PR introduces an issue IMO: the `doSetUp()` and `doTearDown()` don't include the `: void` return type in the "PHPUnit <8 compatible" version of the class. Since Symfony 4 is compatible with PHP 7.1 only, I added the return type in all versions of the `TestCaseSetUpTearDownTrait` to be sure to not trigger the same issues than the one we have with PHPUnit8's BC break.

Pierstoval commented:
The PR is ready, still waiting for reviews 😀

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php:
```diff
@@ -0,0 +1,382 @@
+        Assert::assertSame(
+            $expectedCode,
+            $code = static::getResponse()->getStatusCode(),
+            "Response code [$expectedCode] does not match actual HTTP code [$code]."
```
We are using `""` instead of `[]` to quote things in Symfony.

Pierstoval replied:
Thanks, fixed

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php:
```diff
@@ -0,0 +1,382 @@
+        if (!static::$client instanceof Client) {
+            static::fail(\sprintf(
+                'A client must be set to make assertions on it. Did you forget to call "%s::%s"?',
+                static::class, 'createClient'
```
createClient can be part of the string instead of an argument to `sprintf`

Pierstoval replied:
Thanks for the catch, fixed

fabpot reviewed (commented):
Looks interesting. I have not reviewed all assertions but I've made some cosmetic recommendations.

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php:
```diff
@@ -0,0 +1,382 @@
+
+        Assert::assertTrue(
+            $response->isSuccessful(),
+            "Response was expected to be successful, but actual HTTP code is {$response->getStatusCode()}."
```
Idiomatic Symfony would be to use `sprintf()` here, like `sprintf('Response was expected to be successful, but actual HTTP code is "%s".', $response->getStatusCode())`. That would mean changing the other cases as well.

Pierstoval replied:
I will do this 👍

Pierstoval replied:
Done

Pierstoval commented:
Thanks @fabpot for the review, I fixed the cosmetic things.

fabbot.io's error seems like a false positive to me, since the phpdoc is necessary to sort of "warn" about the fact that the return type is void with PHPUnit 8.

I squashed everything, so the PR seems finished to me, unless anyone has another comment or idea I could work on 😄

fabpot commented on src/Symfony/Component/Form/Test/TestCaseSetUpTearDownTrait.php:
```diff
@@ -45,17 +45,11 @@ protected function tearDown(): void
-         * @return void
-         */
-        private function doSetUp()
+        private function doSetUp(): void
```
Why this change and the one on the Validator component? They don't seem related to your PR.

Pierstoval replied:
I squashed the commit with all the others, but this is intended to fix the fact that PHPUnit 8 needs the type-hint, and not PHPUnit 7.

As Symfony needs PHP 7.1, the void return type can be added with no harm, and since these traits are new as of the upcoming v4.3, IMO they can benefit from this return type right from the start.

I may separate the fix into another commit or into another PR if needed, but since I also need to add such trait in the new `WebTestAssertions`, I thought it would be also necessary to bring the fix to other traits at the same place. 

WYDT? Keep it as-is? New commit/PR?

nicolas-grekas replied:
For 4.2 and should be applied to all similar traits if there are more similar cases.

Pierstoval replied:
So I should split the change and put it in another PR targeted on 4.2?

fabpot replied:
So, let's remove this change from here and let's create another PR for 4.2.

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php:
```diff
@@ -0,0 +1,383 @@
+        Assert::assertSame(
+            $expectedCode,
+            $code = static::getResponse()->getStatusCode(),
+            \sprintf('Response code "%s" does not match actual HTTP code "%s".', $expectedCode, $code)
```
We don't need the `\\` before `sprintf`, we are only using it when it makes a performance difference, which is not the case here.

Pierstoval replied:
Done.

Is there any doc with the list of functions that are concerned by a performance boost with this backslash?

nicolas-grekas replied:
Here is the issue that references them: https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/3048
The process is the following: never add any, then let php-cs-fixer/fabbot add them where needed.

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php:
```diff
@@ -0,0 +1,383 @@
+
+        Assert::assertEquals(
+            $expectedTitle,
+            \trim($title = $titleNode->text()),
```
no leading `\\`

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Test/WebTestAssertions.php:
```diff
@@ -0,0 +1,383 @@
+    {
+        $cookies = static::getResponse()->headers->getCookies();
+
+        $filteredCookies = \array_filter($cookies, function (Cookie $cookie) use ($name, $path, $domain) {
```
Please remove all leading backslashes.

fabpot commented:
@Pierstoval I think we also need some tests for these new assertions.

Pierstoval commented:
@fabpot I'd be interested to know how I could test this, maybe I need a bit of coaching on this. I can create a single test and you could tell me if the direction is good or not 🙂

fabpot commented:
@Pierstoval I've taken over this PR in #30813 where I've refactored the code and added tests.

fabpot commented:
closing in favor of #30813
